### ISSUE TITLE
Are likes likely?

### ISSUE THREAD

**Issue by felixplesoianu** [jaywink/socialhome#206]
It would be nice to have likes, favorites, bookmarks or some such, unless it would conflict with the core concepts of Socialhome. Preferably private, but that's not very important.

**Comment by jaywink** [comment]
Actually, both likes and reshares *will* be coming to Socialhome. Likes will either be the traditional "like" or a more varied "reaction" type. The reaction could be a varied type of emoji basically, a bit like on Slack and in a more limited manner Github and Facebook. Probably start with the traditional "boring" like. About the other idea, see #104.

Thanks for reporting!

**Comment by christophehenry** [comment]
I think it may be a good thing to settle down and start thinking carefully about reshares before implementing them. I don't want SH to become the stream-full-of-reshares mess than pollutes diaspora. This raises two points:

1. all reshares of the same post must be disaplayed *only once* in the stream and share the same comments threads,
2. though, sometimes user may want to create a whole new post with its own comments stream and append her own thoughts to the original post.

At the time, I proposed two distinct type of reshares for diaspora: *spread* which does the first, *reshare* which does the second. But I'm not fond of this idea. Stupidely complex. So here is another proposition: the reshare button opens a popup with the possibility to append something to the original post. If the textarea is empty, app just spreads (displayed once, unified comment stream) otherwise, it reshares (brand new post with a link to the original).

**Comment by jaywink** [comment]
I would really like to unify the reshares instead of showing them separately like Diaspora does, so more like for example Twitter which just shows a counter how many people have retweeted. So in theory a big +++ for this.

The problem that comes from here is that Diaspora protocol allows replying to the reshare directly. What do we do with the comments that target the reshare? We can show them in the original maybe. But would that be confusing? Some kind of visual hint would be required to show that "this comment is replying to a reshare of xxx". Threaded comments (technically already possible) would probably have to be supported so that users can reply to the reshare comments.

I would prefer it this way but the UX could be tricky.

The other option is to not care too much and just do it and don't care where the comments are routed inside Diaspora remotes. ActivityPub, for example, does not AFAICT support replying to the reshare. In AP a "reshare" is actually an activity of type `Announce` (confusing, yes) that targets the original object. It isn't an object itself like in Diaspora. And I don't really want to build SH just modelling everything Diaspora does.

**Comment by felixplesoianu** [comment]
Then do it Fediverse-style. Mastodon intentionally doesn't implement subtweeting, to limit the bullying and trolling it enables on Twitter. And the equivalent functionality on Tumblr (adding content at the bottom of a post you reblog) is also known to cause a lot of grief.

Making reshares just point at the original post gives credit where credit is due, should be easier to implement, *and* easier on the network. Can't see any downsides.

As for the fear that one's followers might just reshare a lot of content you don't care about, there could always be an option not to see reshares. Though by definition I only reblog/retweet/boost only content I would have liked to post myself, OR that I know at least some of my followers would like to see. How else?

**Comment by christophehenry** [comment]
> The problem that comes from here is that Diaspora protocol allows replying to the reshare directly.

Ah... Haven't foreseen that... Maybe a comment carousel displaying the comments from the multiple reshares?

**Comment by jaywink** [comment]
In any case, before doing anything, reshare support needs to be added to the federation library. Logged jaywink/federation#94 for that.

After thinking about it, I think the way to go is to not show reshares separately like Diaspora does. We'll figure out something regarding the comments when that time goes 👍 This does introduce a bit of complexity, for example need to monitor when a user sees a post for the first time. I'd like it to work a bit like Twitter where you only see it once, whether it be the original or someone resharing it. I guess Mastodon works like this too (thus the name "boost")?

I think as far as implementation goes, saving the "seen" in cache for a few weeks might be enough. It wouldn't be too bad to see the same post after a few weeks if someone happens to reshare it. I don't really think it's a good idea to start saving "seen" status of every single post in the database for every single local user. A cache is cheaper to write/read when browsing the stream anyway.

**Comment by christophehenry** [comment]
> I think as far as implementation goes, saving the "seen" in cache for a few weeks might be enough. It wouldn't be too bad to see the same post after a few weeks if someone happens to reshare it. I don't really think it's a good idea to start saving "seen" status of every single post in the database for every single local user. A cache is cheaper to write/read when browsing the stream anyway.

Just make a diff between the creation date and the most recent reshare. If the diff is more than two weeks, for instance, you stop rolling it up.

**Comment by jaywink** [comment]
Some very initial support for shares has been merged in (though not yet deployed to socialhome.network). Remote shares are now saved and a counter is shown in the UI.

Missing steps:
* [ ] Add "share" method to Content API
* [ ] Allow user to share a content item by clicking the share icon
* [ ] Federate shares outwards to remote nodes
* [ ] Do a websocket push on a new share - even if we don't have client code doing anything with that yet
* [ ] Upstream federation library method to fetch unknown shared content when receiving a share
* [ ] Ensure visibility constraints are respected when processing a remote share. A share can't have public visibility if the original content is limited. Concerns mostly SH nodes since diaspora only has public shares.
* [ ] Ensure shares are retracted properly towards remotes when they are deleted.
* [ ] "Promote" content in streams according to shares to users who would not otherwise see it (not following original author, etc). Must keep some kind of flag whether user has seen the content.

Things to do later (will log separate issues)
* [ ] Allow sharing a reply .. this is already possible in the database, so why not! We can only federate this to other SH nodes though. UI should do some kind of "reply highlight" which has a link to the reply parent content.
* [ ] Send participation on share to remotes if it was a local content that was shared. This mimics how Diaspora works internally, but requires participation support in the federation library.
* [ ] Allow "quoting" a share. This will only federate properly to other SH nodes.
